File: PA43.java
//Leetcode 532. K-diff Pairs in an Array
//Time Complexity : O(nlogn) // for soting O(nlogn) and O(n) for iterating over array nums 
//Space Complexity: O(n) // n is length of nums

class Solution {
    public int findPairs(int[] nums, int k) {
        
        int count =0;
        if(nums==null || nums.length<2 || k<0){
            return count;
        }
        Arrays.sort(nums);
        HashSet<Integer> hs= new HashSet<>();       
        int fl=0;
        if(k==0){            
            for(int i: nums){
                if(hs.contains(i)){
                   if(fl<1) {
                    fl++;
                    count++;
                   }
                }
                else{
                    fl=0;
                    hs.add(i);
                }
            }
            return count;
        }

                for(int i=0; i<nums.length;i++)
            {
                if(!hs.contains(nums[i]) && hs.contains(nums[i]-k)){

                    count++;
                }

                    hs.add(nums[i]);

            }          
        return count;        
    }
}

File: PA_43.py
#LC 532 - K-diff Pairs in an Array
#Time complexity - O(nlogn)
#Space Complexity - O(1)
class Solution(object):
    def findPairs(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        if not nums:
            return 0
        nums = sorted(nums)
        i,j = 0,1
        ans = 0
        while j < len(nums) and i < len(nums):
            if abs(nums[j] - nums[i]) == k and i != j:
                #print(j,i)
                ans = ans + 1 
                while i < len(nums)-1 and nums[i] == nums[i+1]:
                    i = i + 1
                while j < len(nums)-1 and nums[j] == nums[j+1]:
                    j = j + 1
                j = j + 1
                i = i + 1
            elif abs(nums[j] - nums[i]) > k:
                i = i + 1
            else:
                j = j + 1
        return(ans)
            